pykokkos/interface/type_inference.py

```diff
@@ -0,0 +1,234 @@
+from .views import View, ViewType
+from .layout import Layout
+from .data_types import DataType, DataTypeClass
+
+@dataclass
+class HandledArgs:
+    """
+    Class for holding the arguments passed to parallel_* functions
+    """
+
+    name: Optional[str]
+    policy: ExecutionPolicy
+    workunit: Callable
+    view: Optional[ViewType]
+    initial_value: Union[int, float]
+
+
+@dataclass
+class UpdatedTypes:
+    """
+    Class for storing inferred type annotation information 
+    (Making Pykokkos more pythonic by automatically inferring types)
+    """
+
+    workunit: Callable
+    inferred_types: Dict[str, str] # type information stored as string: identifier -> type
+    is_arg: set[str]
+    layout_change: Dict[str, str] # if layout for a view is not Layout.Default it will be stored here
+
+
+def handle_args(is_for: bool, *args) -> HandledArgs:
+    """
+    Handle the *args passed to parallel_* functions
+
+    :param is_for: whether the arguments belong to a parallel_for call
+    :param *args: the list of arguments being checked
+    :returns: a HandledArgs object containing the passed arguments
+    """
+
+    unpacked: Tuple = tuple(*args)
+
+    name: Optional[str] = None
+    policy: Union[ExecutionPolicy, int]
+    workunit: Callable
+    view: Optional[ViewType] = None
+    initial_value: Union[int, float] = 0
+
+
+    if len(unpacked) == 2:
+        policy = unpacked[0]
+        workunit = unpacked[1]
+
+    elif len(unpacked) == 3:
+        if isinstance(unpacked[0], str):
+            name = unpacked[0]
+            policy = unpacked[1]
+            workunit = unpacked[2]
+        elif is_for and isinstance(unpacked[2], ViewType):
+            policy = unpacked[0]
+            workunit = unpacked[1]
+            view = unpacked[2]
+        elif isinstance(unpacked[2], (int, float)):
+            policy = unpacked[0]
+            workunit = unpacked[1]
+            initial_value = unpacked[2]
+        else:
+            raise TypeError(f"ERROR: wrong arguments {unpacked}")
+
+    elif len(unpacked) == 4:
+        if isinstance(unpacked[0], str):
+            name = unpacked[0]
+            policy = unpacked[1]
+            workunit = unpacked[2]
+
+            if is_for and isinstance(unpacked[3], ViewType):
+                view = unpacked[3]
+            elif isinstance(unpacked[3], (int, float)):
+                initial_value = unpacked[3]
+            else:
+                raise TypeError(f"ERROR: wrong arguments {unpacked}")
+        else:
+            raise TypeError(f"ERROR: wrong arguments {unpacked}")
+
+    else:
+        raise ValueError(f"ERROR: incorrect number of arguments {len(unpacked)}")
+
+    if isinstance(policy, int):
+        policy = RangePolicy(km.get_default_space(), 0, policy)
+
+    return HandledArgs(name, policy, workunit, view, initial_value)
+
+
+def get_annotations(parallel_type: str, handled_args: HandledArgs, *args, passed_kwargs) -> UpdatedTypes:
```

Comment:
If this is meant to return `None` in some cases, the return type should be `Optional[UpdatedTypes]`